architecture-decision-records/0002-versioning-and-deployments.md

```diff
@@ -0,0 +1,75 @@
+# 2. Versioning and deployments
+
+**Date:** 2020-02-21
+
+## Context
+
+TDR will have multiple deployment environments: integration, staging and
+production. We want to be able to deploy new versions of our services to each
+environment in turn.
+
+## Decision
+
+Configure each project's build and Jenkinsfile so that when code is merged to
+master and the tests pass, Jenkins runs the following steps:
+
+1. Create a git tag with a version number based on the Jenkins build number, at
+   the commit that master is pointing to
+2. Push the version tag to GitHub
+3. If the repo is deployed as a Docker image (such as the [consignment
+   API][tdr-api] or the [frontend]), build the Docker image, tag it
+   with the same version number and push it to Docker Hub
+
+This version is now available to deployed to each environment in turn.
+
+The master build should start an automated deployment to the Integration
+environment, passing the version tag as a parameter. The deployment build will:
+
+1. Pull the Docker image tagged with the given version tag
+2. Retag the Docker image with the environment name (e.g. "intg")
+3. Trigger an update in AWS, e.g. by calling the AWS API to restart the ECS
+   service, which will pull the environment-specific tag
+4. Push the git release branch matching the environment name to the same commit
+   as the given version tag
+
+A developer can then manually start a release to Staging then Production,
+passing the same version tag as a parameter OR the "intg" branch/tag
+name.
+
+[tdr-api]: https://github.com/nationalarchives/tdr-consignment-api
+[frontend]: https://github.com/nationalarchives/tdr-transfer-frontend
+
+### Version numbers
+
+The version numbers will come from Jenkins build numbers, rather than using a
+semantic versioning system. This is because semantic versions aren't as
+meaningful for applications as for libraries, and because assigning a semantic
+version would need a developer to bump the version number. It's more useful to
+have an auto-incrementing version number generated by something like Jenkins.
```

Comment:
We wouldn't strictly need a developer to bump the version number, the generated code libraries and the graphql client all have auto version bumped versions.
The point about not using SemVer stands though, the build numbers are fine, it's just needing developers to bump the version is not a reason to not use it if you see what I mean.

Reply:
I've updated it to say that a minor or major version bump would need a person to do it. I take your point that you _can_ update the version number automatically, but it might be confusing because it looks like a semver number but it isn't really one.